src/app/pages/system/view-enclosure/components/jbof-list/jbof-list.component.ts

```diff
@@ -0,0 +1,133 @@
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class JbofListComponent implements OnInit {
+  protected readonly requiredRoles = [Role.JbofWrite];
+
+  filterString = '';
+  jbofs: Jbof[] = [];
+  isJbofLicensed = false;
+
+  dataProvider: AsyncDataProvider<Jbof>;
+  columns = createTable<Jbof>([
+    textColumn({
+      title: this.translate.instant('Description'),
+      propertyName: 'description',
+      sortable: true,
+    }),
+    textColumn({
+      title: this.translate.instant('IPs'),
+      getValue: (row) => [row.mgmt_ip1, row.mgmt_ip2].filter(Boolean).join(', '),
+    }),
+    textColumn({
+      title: this.translate.instant('Username'),
+      propertyName: 'mgmt_username',
+      sortable: true,
+    }),
+    actionsColumn({
+      actions: [
+        {
+          iconName: 'edit',
+          tooltip: this.translate.instant('Edit'),
+          onClick: (row) => this.openForm(row),
+        },
+        {
+          iconName: 'delete',
+          tooltip: this.translate.instant('Delete'),
+          requiredRoles: this.requiredRoles,
+          onClick: (row) => this.doDelete(row),
+        },
+      ],
+    }),
+  ], {
+    rowTestId: (row) => 'jbof-' + row.mgmt_username,
+  });
+
+  get emptyConfigService(): EmptyService {
+    return this.emptyService;
+  }
+
+  constructor(
+    private ws: WebSocketService,
+    private slideInService: IxSlideInService,
+    private dialogService: DialogService,
+    private errorHandler: ErrorHandlerService,
+    private translate: TranslateService,
+    private emptyService: EmptyService,
+    private loader: AppLoaderService,
+    private cdr: ChangeDetectorRef,
+  ) { }
+
+  ngOnInit(): void {
+    const request$ = this.ws.call('jbof.query').pipe(
+      tap((jbofs) => this.jbofs = jbofs),
+      untilDestroyed(this),
+    );
+    this.dataProvider = new AsyncDataProvider(request$);
+    this.getJbofs();
+  }
+
+  openForm(jbof?: Jbof): void {
+    const slideInRef = this.slideInService.open(JbosFormComponent, { data: jbof });
+    slideInRef.slideInClosed$.pipe(untilDestroyed(this)).subscribe(() => this.getJbofs());
+  }
+
+  doDelete(jbof: Jbof): void {
+    this.dialogService.confirm({
+      title: this.translate.instant('Delete'),
+      message: this.translate.instant('Are you sure you want to delete this item?'),
+      buttonText: this.translate.instant('Delete'),
+    }).pipe(
+      filter(Boolean),
+      switchMap(() => this.ws.call('jbof.delete', [jbof.id]).pipe(this.loader.withLoader())),
+      untilDestroyed(this),
+    ).subscribe({
+      next: () => this.getJbofs(),
+      error: (error: unknown) => {
+        this.dialogService.error(this.errorHandler.parseError(error));
+      },
+    });
+  }
+
+  getJbofs(): void {
+    this.dataProvider.load();
+    this.updateJbofLicensed();
+  }
+
+  updateJbofLicensed(): void {
+    this.ws.call('jbof.licensed').pipe(untilDestroyed(this)).subscribe((licensed) => {
+      this.isJbofLicensed = licensed > 0;
```

Comment:
We should also track number of shelves added vs number of shelves licensed and disable an Add button with a tooltip when this is the case.